### PR TITLE
Add support for macOS brew

### PR DESCRIPTION
**Pull request by tuedachu**
First version of 'helm-system-package' for home brew 

Working functionalities:
- Show information
- Install/Uninstall
- Browse URL
- Link/Unlink app

TODO:
(1) Implement `show-dependencies' function
(2) Show packages explicitly installed (function `helm-system-packages-brew-list-explicit' already exists)
(3) Implement `find-files'
(4) Keymap

### CODE REVIEW

**Comment by Ambrevar** [comment]
Thank you very much for this contribution!

Before we can proceed, please rebase your pull request onto upstream master and make sure only significant commits present.
While it's good practice to make commits atomic, it's OK if the `-brew.el` file is provided through only one commit since the file is new.  That is to say, unless some commits are particularly relevant (pitfalls, exceptions, special cases, etc.).  Use your better judgment.

File: helm-system-packages-brew.el [hunk 1]
```diff
@@ -0,0 +1,214 @@
+;;; helm-system-packages-brew.el --- Helm UI for Mac OS ' homebrew. -*- lexical-binding: t -*-
```
Comment (Ambrevar) [inline helm-system-packages-brew.el]: The OS name is officially "macOS".  Also no space before the apostrophe.

File: helm-system-packages-brew.el [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+;;; Commentary:
+;; Helm UI for Mac OS' homebrew.
```
Comment (Ambrevar) [inline helm-system-packages-brew.el]: "macOS".

File: helm-system-packages-brew.el [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    (setq names
+	  (replace-regexp-in-string ":.*" "" descriptions))      
+    (setq descriptions (mapconcat (lambda (package-from-list)
+				    (let* ((pkg (split-string package-from-list ": "))
+					   (name (car pkg))
+					   (desc (car (cdr pkg)))
+					   (format-string (format "%%-%ds  %%s" helm-system-packages-column-width)))
+				    (format format-string name desc)))
+				  (split-string descriptions"\n") "\n" ))
+  (cons names descriptions)))
```
Comment (Ambrevar) [inline helm-system-packages-brew.el]: The cache construction is a little different now, you should use `helm-system-packages--cache-set`.  See -pacman.el or -xbps.el.

File: helm-system-packages-brew.el [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+  (unless (and helm-system-packages--names helm-system-packages--descriptions)
+    (helm-system-packages-brew-refresh))
+  ;; TODO: We should only create the buffer if it does not already exist.
+  ;; On the other hand, we need to be able to override the package list.
+  ;; (unless (helm-candidate-buffer) ...
+  (helm-init-candidates-in-buffer
+      'global
+    (if helm-system-packages-show-descriptions-p
+        helm-system-packages--descriptions
+      helm-system-packages--names)))
```
Comment (Ambrevar) [inline helm-system-packages-brew.el]: This function can now be removed altogether: use `helm-system-packages-init` instead.

File: helm-system-packages-brew.el [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+(defun helm-system-packages-brew-refresh ()
+  "Refresh the package list."
+  (interactive)
+  (setq helm-system-packages--source-name "brew source")
```
Comment (Ambrevar) [inline helm-system-packages-brew.el:81]: The title is now stored in the cache.  See `helm-system-packages--cache`.

File: helm-system-packages-brew.el [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+  (setq helm-system-packages-column-width
+        (or helm-system-packages-brew-column-width
+            helm-system-packages-column-width))
+    (let ((res (helm-system-packages-brew-cache)))
+      (setq helm-system-packages--names (car res)
+            helm-system-packages--descriptions (cdr res)))
+    (setq helm-system-packages--display-lists nil))
```
Comment (Ambrevar) [inline helm-system-packages-brew.el:88]: Display lists are important: they allow for fontifying the different categories (installed packages, dependencies, orphans, etc.).  Again, see `helm-system-packages-(xbps|pacman)-refresh`.

File: helm-system-packages-brew.el [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+      helm-system-packages--names)))
+
+(defun helm-system-packages-brew-refresh ()
+  "Refresh the package list."
+  (interactive)
+  (setq helm-system-packages--source-name "brew source")
+  (setq helm-system-packages-column-width
+        (or helm-system-packages-brew-column-width
+            helm-system-packages-column-width))
+    (let ((res (helm-system-packages-brew-cache)))
```
Comment (Ambrevar) [inline helm-system-packages-brew.el:85]: The caching mechanism has changed, as stated previously.  Refer to the refresh functions of XBPS or pacman for an example.

File: helm-system-packages-brew.el [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+Otherwise display in `helm-system-packages-buffer'."
+  (let* ((descriptions (json-read-from-string (with-temp-buffer
+						(apply 'call-process "brew" nil t nil "info" "--json=v1" (helm-marked-candidates)) 
+						(buffer-string))))
+	 desc-list
+	 pkg-desc-alist
+	 str
+	 pkg
+	 (i 0))
+    (dolist (pkg (helm-marked-candidates) nil)
```
Comment (Ambrevar) [inline helm-system-packages-brew.el]: `nil` is unnecessary here.

File: helm-system-packages-brew.el [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+	      "* Dependencies:\n" "   " (mapconcat 'identity (alist-get 'dependencies pkg-desc-alist) "\n   ") "\n\n"
+	      "* Optional dependencies:\n" "   "(mapconcat 'identity (alist-get 'optional_dependencies pkg-desc-alist) "\n   ") "\n\n"
+	      "* Options:\n" (mapconcat (lambda (pkg-option)
+					(concat (alist-get 'option pkg-option)  "\n"
+						"    " (alist-get 'description pkg-option)  "\n"))
+				      (alist-get 'options pkg-desc-alist) "\n")
+	      "\n\n"
+	      "* Caveats: " (alist-get 'caveats pkg-desc-alist) "\n"))
+     (add-to-list 'desc-list `(uninstalled (,pkg . ,str)))
+     (setq i (+ i 1)))
```
Comment (Ambrevar) [inline helm-system-packages-brew.el]: Use the `1+` function.

File: helm-system-packages-brew.el [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+    (helm-system-packages-browse-url (mapcar (lambda(pkg)
+					       (alist-get 'homepage pkg))
+					     descriptions))))
+
+(defun helm-system-package-brew-link-app (_candidate)
+  (helm-system-packages-brew-run  "brew" "link"))
+
+(defun helm-system-package-brew-unlink-app (_candidate)
+  (helm-system-packages-brew-run  "brew" "unlink"))
+
```
Comment (Ambrevar) [inline helm-system-packages-brew.el:132]: No need for multiple line breaks, I think.

File: helm-system-packages-brew.el [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+	      "* Caveats: " (alist-get 'caveats pkg-desc-alist) "\n"))
+     (add-to-list 'desc-list `(uninstalled (,pkg . ,str)))
+     (setq i (+ i 1)))
+    (helm-system-packages-show-information desc-list)))
+
+(defun helm-system-package-brew-browse-url (_candidate)
+   (let* ((descriptions (json-read-from-string (with-temp-buffer
+						(apply 'call-process "brew" nil t nil "info" "--json=v1" (helm-marked-candidates)) 
+						(buffer-string)))))
+    (helm-system-packages-browse-url (mapcar (lambda(pkg)
```
Comment (Ambrevar) [inline helm-system-packages-brew.el]: Missing space before `(pkg)`.

File: helm-system-packages-brew.el [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+(defun helm-system-package-brew-link-app (_candidate)
+  (helm-system-packages-brew-run  "brew" "link"))
+
+(defun helm-system-package-brew-unlink-app (_candidate)
+  (helm-system-packages-brew-run  "brew" "unlink"))
+
+
+
+(defun helm-system-packages-brew-run (command &rest args)
```
Comment (Ambrevar) [inline helm-system-packages-brew.el]: Just to be sure: This is the same as `helm-system-packages-call-as-root` except that it does not use `sudo`, right?
If so, mention it in the documentation.  Also `helm-system-packages-call-as-root` has undergone some changes, you might want to reflect them here.

File: helm-system-packages-brew.el [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+    ("Install (`C-u' to reinstall)" .
+     (lambda (_)
+       (if helm-current-prefix-arg
+	   (helm-system-packages-brew-run  "brew" "reinstall")
+	 (helm-system-packages-brew-run "brew" "install"))))
+    ("Uninstall (`C-u' to uninstall all versions)" .
+     (lambda (_)
+       (helm-system-packages-brew-run "brew" "uninstall"
+                                         (when helm-current-prefix-arg "--force"))))
+
```
Comment (Ambrevar) [inline helm-system-packages-brew.el]: No linebreak here.

File: helm-system-packages-brew.el [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+    ("Browse homepage URL" . helm-system-package-brew-browse-url)
+    ("Link application" . helm-system-package-brew-link-app)
+    ("Unlink application" . helm-system-package-brew-unlink-app))
+  "Actions for Helm brew."
+  :group 'helm-system-packages
+  :type '(alist :key-type string :value-type function))
+
+(defun helm-system-packages-brew-build-source ()
+  "Build Helm source for brew."
+  (helm-build-in-buffer-source helm-system-packages--source-name
```
Comment (Ambrevar) [inline helm-system-packages-brew.el]: The way the source name is set has changed: see -xbps.el or -pacman.el.

**Review by Ambrevar — COMMENTED** [review]
Looks good overall but your contribution went out of sync with master since I contributed the support for remote systems via TRAMP.  Bad timing, sorry!
Try to address my comments, test and repush, I'll merge then.

Let me know if you run into any issue.  Feel free to ask for further explanations regarding the new cache system.

**Comment by piotr-yuxuan** [comment]
I've tried to helped for the review process by addressing your most simple remarks, however I'm unsure about where I should put these changes to be helpful. I've opened PR https://github.com/tuedachu/helm-system-packages/pull/1 to @tuedachu's PR, is it a good pattern? is there anything simpler? Also, I commented this PR to make clear which remarks of yours are still to be addressed and which ones should be fixed now. Just tell me if you consider it's useless noise or if it can smoothen your review process!

File: helm-system-packages-brew.el (hunk 1 above)
Comment (piotr-yuxuan) [reply helm-system-packages-brew.el]: Should be fixed by https://github.com/tuedachu/helm-system-packages/pull/1.

File: helm-system-packages-brew.el (hunk 2 above)
Comment (piotr-yuxuan) [reply helm-system-packages-brew.el]: Should be fixed by https://github.com/tuedachu/helm-system-packages/pull/1.

File: helm-system-packages-brew.el (hunk 3 above)
Comment (piotr-yuxuan) [reply helm-system-packages-brew.el]: Should be fixed by https://github.com/tuedachu/helm-system-packages/pull/1.

File: helm-system-packages-brew.el (hunk 10 above)
Comment (piotr-yuxuan) [reply helm-system-packages-brew.el:132]: Should be fixed by https://github.com/tuedachu/helm-system-packages/pull/1.

File: helm-system-packages-brew.el (hunk 11 above)
Comment (piotr-yuxuan) [reply helm-system-packages-brew.el]: Should be fixed by https://github.com/tuedachu/helm-system-packages/pull/1.

File: helm-system-packages-brew.el (hunk 13 above)
Comment (piotr-yuxuan) [reply helm-system-packages-brew.el]: Should be fixed by https://github.com/tuedachu/helm-system-packages/pull/1.

**Comment by tuedachu** [comment]
@piotr-yuxuan,
Thanks for your contribution. I have been quite busy for the last weeks, and I did not have time to review @Ambrevar 's comments. I wil try to work on it tomorrow (to make it compatible with the current version of 'helm-system-packages'). I am planning to make a new PR by the end of this week.

Note that many functionalities are still missing, so your contribution can be very useful. I am not a very experienced user of brew, I just covered the basic functionalities (instal, uninstal, linkapp, show information).

**Comment by piotr-yuxuan** [comment]
If I have time, I'll try to look into this.

**Comment by Ambrevar** [comment]
I've just pushed initial support for Guix.  Since it is slightly different from the other package managers, you might find it interesting to study.

**Comment by tuedachu** [comment]
@Ambrevar ,
I finalized the first version of helm-system-packages for macOS brew. Basic functionalities work fine.
Any comment is of course very welcome.

@piotr-yuxuan ,
Please feel free to contribute. I will sell my mac soon, so I won't be able to continue developing this software.

**Comment by Ambrevar** [comment]
Thank you very much for your work!

**Pull request merged**